# Overview

This repository houses a Snakemake pipeline that maps read sets in FASTQ format to a reference assembly in FASTA format, then generates two consensus sequences using `bcftools consensus` and `racon`. Diagnostic plots are then generated by re-mapping the reads to the consensus sequences.

# Installation

Begin by cloning and cd-ing into the directory using:

`git clone https://github.com/Danderson123/SBE_180391 && cd SBE_180391`

## Conda

The required dependencies can be installed through conda using:

`conda env create -f environment.yaml`

or if you have [mamba](https://mamba.readthedocs.io/en/latest/installation/mamba-installation.html) installed:

`mamba env create -f environment.yaml`

## Singularity (preferred method)

This repository includes a recipe and pre-built [singularity]() container. The pre-built container is available with each release and the container can be built with:

`sudo singularity build SBE_180391.img Singularity.def`

# Configuration

The reference genome (in FASTA) format used to generate the consensus can be changed by modifying the value for `reference_FASTA` in `config.yaml` to the absolute path of the new reference FASTA.

The directory containing the long reads to be analysed can be modify by changing the value for `nanopore_directory` in `config.yaml` to the absolute path to the new directory.

The directory containing the short reads to be analysed can be modify by changing the value for `illumina_directory` in `config.yaml` to the absolute path to the new directory **NOTE: Illumina reads are assumed to be paired-end and deinterleaved**.
